### ISSUE TITLE
How can I make my own loss defination

### ISSUE THREAD

**Issue by Kaido0** [jocicmarko/ultrasound-nerve-segmentation#26]
I have my own training sets,but the background are too large~Also,I know the background is black(pixel=0),When I am caculating the dice,I want to 'remove' the background(Don't caculate it into dice).
I try this:for i in y_true_f.eval():
But it didn't work~

**Comment by jocicmarko** [comment]
In current implementation "black" pixels are not calculated into the Dice score, only "white" ones.

**Comment by Kaido0** [comment]
@jocicmarko Oh!You are so great~
I saw the new code,but I find Dice code not changed.Where did you add code to judge the pixel? Is it caculated by Keras backend?
Thank you so much!I learn keras all by your code.Haha~

**Comment by jocicmarko** [comment]
@Kaido0 
> Where did you add code to judge the pixel?

If you take a look at the Dice loss function, you will notice that part of the code
`intersection = K.sum(y_true_f * y_pred_f)` does multiplication of true pixels with predicted pixels. Since 1s are for mask 0s are for background, so some of the predicted values are multiplied by 0s in true masks, thus "ignoring" the background! Hope this makes it clearer.

**Comment by Kaido0** [comment]
@jocicmarko Thanks~I got it.
I made a mistake,I want to divide the image into four categories(Include background).So I need to  make "0" pixels are not calculated into the Dice score, only '1','0.25','0.75' .I want to use'softmax' and` conv10 = Convolution2D(4, 1, 1, activation='softmax')(conv9)`,but I think it is not clearly,Would you give me some suggestion,thank you very much

**Comment by EloyRoura** [comment]
@jocicmarko Thanks for your code, it helped me quite a lot to get my unet architecture.
@Kaido0 I run into the same issue with the dice computation and this is what I tried:
def dice_coef(y_true, y_pred):
    y_true_f = K.flatten(y_true[...,1:])    
    y_pred_f = K.flatten(y_pred[...,1:])
    intersection = K.sum(y_true_f * y_pred_f)
    return (2. * intersection + smooth) / (K.sum(y_true_f) + K.sum(y_pred_f) + smooth)

The network runs, but I got weird dice now. Basically I have 4 classes being (background, and 3 more within the image), so my y_true should look like (nb_samples, img_size, 4). I am doing a quick test to see whether this DICE is correct, basically using only one image volume for training and then I predict using the exact same volume so the DICE for every class should be the same. Well, it is not even close.
Anyway, hope this helps you. If you figure that out, I would appreciate a feedback on that.
